File: card.c
#include "card.h"

#include <ctype.h>
#include <stddef.h>
#include <string.h>

static Rank parse_rank_char(char c) {
    c = (char)toupper((unsigned char)c);

    if (c >= '2' && c <= '9') {
        return (Rank)(c - '0');
    }

    switch (c) {
        case 'T': return RANK_T;
        case 'J': return RANK_J;
        case 'Q': return RANK_Q;
        case 'K': return RANK_K;
        case 'A': return RANK_A;
        default: return RANK_INVALID;
    }
}

static Suit parse_suit_char(char c) {
    c = (char)tolower((unsigned char)c);

    switch (c) {
        case 'c': return SUIT_CLUBS;
        case 'd': return SUIT_DIAMONDS;
        case 'h': return SUIT_HEARTS;
        case 's': return SUIT_SPADES;
        default: return SUIT_INVALID;
    }
}

int parse_card(const char *text, Card *card) {
    Rank rank;
    Suit suit;

    if (text == NULL || card == NULL || strlen(text) != 2) {
        return 0;
    }

    rank = parse_rank_char(text[0]);
    suit = parse_suit_char(text[1]);

    if (rank == RANK_INVALID || suit == SUIT_INVALID) {
        return 0;
    }

    card->rank = rank;
    card->suit = suit;
    return 1;
}

int parse_hole_cards(const char *text1, const char *text2, HoleCards *hand) {
    Card card1;
    Card card2;

    if (hand == NULL) {
        return 0;
    }

    if (!parse_card(text1, &card1) || !parse_card(text2, &card2)) {
        return 0;
    }

    if (is_same_card(card1, card2)) {
        return 0;
    }

    hand->card1 = card1;
    hand->card2 = card2;
    return 1;
}

int is_same_card(Card a, Card b) {
    return a.rank == b.rank && a.suit == b.suit;
}

void card_to_string(Card card, char *buffer) {
    static const char ranks[] = "??23456789TJQKA";
    char suit = '?';

    if (buffer == NULL) {
        return;
    }

    switch (card.suit) {
        case SUIT_CLUBS: suit = 'c'; break;
        case SUIT_DIAMONDS: suit = 'd'; break;
        case SUIT_HEARTS: suit = 'h'; break;
        case SUIT_SPADES: suit = 's'; break;
        default: suit = '?'; break;
    }

    if (card.rank >= RANK_2 && card.rank <= RANK_A) {
        buffer[0] = ranks[card.rank];
    } else {
        buffer[0] = '?';
    }
    buffer[1] = suit;
    buffer[2] = '\0';
}

File: card.py
from __future__ import annotations

from common import Card, HoleCards, Rank, Suit

BOARD_SIZE = 5
DECK_SIZE = 52
FULL_DECK = tuple(
    Card(rank=Rank(rank), suit=Suit(suit))
    for suit in range(int(Suit.CLUBS), int(Suit.SPADES) + 1)
    for rank in range(int(Rank.TWO), int(Rank.ACE) + 1)
)


_RANK_CHARS = {
    "2": Rank.TWO,
    "3": Rank.THREE,
    "4": Rank.FOUR,
    "5": Rank.FIVE,
    "6": Rank.SIX,
    "7": Rank.SEVEN,
    "8": Rank.EIGHT,
    "9": Rank.NINE,
    "T": Rank.TEN,
    "J": Rank.JACK,
    "Q": Rank.QUEEN,
    "K": Rank.KING,
    "A": Rank.ACE,
}

_SUIT_CHARS = {
    "c": Suit.CLUBS,
    "d": Suit.DIAMONDS,
    "h": Suit.HEARTS,
    "s": Suit.SPADES,
}

_RANK_TO_CHAR = {
    Rank.TWO: "2",
    Rank.THREE: "3",
    Rank.FOUR: "4",
    Rank.FIVE: "5",
    Rank.SIX: "6",
    Rank.SEVEN: "7",
    Rank.EIGHT: "8",
    Rank.NINE: "9",
    Rank.TEN: "T",
    Rank.JACK: "J",
    Rank.QUEEN: "Q",
    Rank.KING: "K",
    Rank.ACE: "A",
}

_SUIT_TO_CHAR = {
    Suit.CLUBS: "c",
    Suit.DIAMONDS: "d",
    Suit.HEARTS: "h",
    Suit.SPADES: "s",
}


def parse_card(text: str) -> Card | None:
    if text is None or len(text) != 2:
        return None

    rank = _RANK_CHARS.get(text[0].upper())
    suit = _SUIT_CHARS.get(text[1].lower())
    if rank is None or suit is None:
        return None

    return Card(rank=rank, suit=suit)


def parse_hole_cards(text1: str, text2: str) -> HoleCards | None:
    card1 = parse_card(text1)
    card2 = parse_card(text2)

    if card1 is None or card2 is None:
        return None
    if is_same_card(card1, card2):
        return None

    return HoleCards(card1=card1, card2=card2)


def is_same_card(a: Card, b: Card) -> bool:
    return a.rank == b.rank and a.suit == b.suit


def card_to_string(card: Card) -> str:
    return f"{_RANK_TO_CHAR.get(card.rank, '?')}{_SUIT_TO_CHAR.get(card.suit, '?')}"
